torch/_inductor/codecache.py

```diff
@@ -575,79 +588,212 @@ def __init__(self, fx_args: List[Any], fx_kwargs: Dict[str, Any]) -> None:
+        """
+        return [s for s in inputs if isinstance(s, torch.SymInt)]
+
+    @staticmethod
+    def _get_shape_env() -> ShapeEnv:
+        """
+        Helper to get the shape env from the tracing context.
+        """
+        tracing_context = torch._guards.TracingContext.get()
+        assert tracing_context is not None
+        return tracing_context.fake_mode.shape_env
+
+    @staticmethod
+    def _lookup_graph(
+        key: str,
+        example_inputs: List[torch.Tensor],
+    ) -> Optional[CompiledFxGraph]:
+        """
+        Lookup a compiled graph in the cache by key. On a hit, return the
+        deserialized CompiledFxGraph object. On a miss, return None.
+        """
+        subdir = FxGraphCache._get_subdir(key)
+        if not os.path.exists(subdir):
+            return None
+
+        # Iterate over any entries in the subdir for this key and evaluate
+        # their guards to determine whether there's a hit.
+        for path in os.listdir(subdir):
+            with open(os.path.join(subdir, path), "rb") as f:
+                graph: CompiledFxGraph = pickle.load(f)
+
+            if not graph.guards_expr:
+                # No guards to evaluate
+                return graph
+
+            # Evaluate the guard expression in the current context.
+            shape_env = FxGraphCache._get_shape_env()
+            symints = FxGraphCache._filter_symints(example_inputs)
+
+            # We don't want to create the guards in the current shape env
+            # unless we have a hit.
+            with shape_env.suppress_guards():
+                res = shape_env.evaluate_guards_expression(graph.guards_expr, symints)
+                hit = definitely_true(res)
+
+            log.debug(
+                "fx graph evaluating guards for key %s: (%s) = %s",
+                key,
+                graph.guards_expr,
+                hit,
+            )
+            if hit:
+                # Now re-evaluate to add the guards to the current shape env.
+                # We have to clear the `evaluate_expr` lru_cache to force evaluation.
+                shape_env.evaluate_expr.cache_clear()
+                shape_env.evaluate_guards_expression(graph.guards_expr, symints)
+                return graph
+
+        return None
+
+    @staticmethod
+    def _save_graph(
+        key: str, compiled_graph: CompiledFxGraph, example_inputs: List[torch.Tensor]
+    ):
+        """
+        Store a serialized CompiledFxGraph on disk.
+        """
         disk_compiled_graph = copy(compiled_graph)
-        # Important as compiled models are not pickleable
-        # TODO: Check status of PR #101651 as that might change the above statement
+        # Important as compiled models are not pickleable:
         disk_compiled_graph.compiled_artifact = None
-        write(pickle.dumps(disk_compiled_graph), "cg", extra=key, hash_type="cg")
 
-    @classmethod
-    def load_graph(cls, cg_path: str) -> CompiledFxGraph:
```

Comment:
The diff looks weird here. `load_graph` is replaced by `lookup_graph` above and the implementation here is for `save_graph`.  So ignore the red lines I guess.